### ISSUE TITLE
Passing Signal as Value or Itself Does Not Trigger Reactive Updates in Child Components

### ISSUE THREAD

**Issue by thetsp** [solidjs/solid#2320]
### Describe the bug

In Solid.js, passing a signal either as its value (count()) or as itself (count) to a child component should allow for reactivity. However, in the following example, neither approach triggers reactivity in the child components (Child and Child2):

import { createSignal } from 'solid-js';

const Even = () => <span>I am even</span>;
const Odd = () => <span>I am odd</span>;

const Child = (props) => {
  // Expected to work when passing count() as value but it's not reactive
  return props.count % 2 === 0 ? <Even /> : <Odd />;
};

const Child2 = (props) => {
  // Expected to work when passing count as signal but it's not updating
  return props.count() % 2 === 0 ? <Even /> : <Odd />;
};

function App() {
  const [count, setCount] = createSignal(0);

  return (
    <div>
      <button onClick={() => setCount(count() + 1)}>{count()}</button>
      <Child count={count()} />
      <Child2 count={count} />
    </div>
  );
}

export default App;

### Your Example Website or App

https://stackblitz.com/edit/solidjs-templates-9kebyj?file=src%2FApp.jsx

### Steps to Reproduce the Bug or Issue

Use the code above.
Click the button to increment the count.
Neither Child nor Child2 reactively update based on the new count value.

### Expected behavior

Child: When passing count() as a value, it should render the correct child (Even or Odd) based on the value of count(). However, this does not work because it's not reactive—Child does not update when count changes.
Child2: When passing the signal count directly, it should reactively update whenever the signal changes. However, the component does not update correctly in this case either.

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by titoBouzout** [comment]
For it to be reactive, it has to be wrapped in jsx, this is how things become reactive in solid
```js
const Child = (props) => {
  return <>{props.count % 2 === 0 ? <Even /> : <Odd />}</>;
};
```

Trying to explain why the example given won't work, for something to be reactive it has to be a function.

If you want to make `Child` reactive, taking `Child` as the `function` to re-run when reactive values change, then that will mean you will be rerunning the full `Child` on each change, which is undesired as it won't be fine-grained.

```js
// this not solid-js code
const Child = (props) => {
  somethingElse() // this will re-run if props.count change
  return props.count % 2 === 0 ? <Even /> : <Odd />;
};
```

You can think of jsx as to be the following, which returns a new function that can be re-run when the count changes.

```js
// this not solid-js code
const Child = (props) => {
  somethingElse()  // this won't re-run when props.count change
  return () => props.count % 2 === 0 ? <Even /> : <Odd />; // this function re-runs when props.count change
};
```

So the idiomatic way of making things reactive in solid is to wrap these in jsx as follows
https://playground.solidjs.com/anonymous/672bae4b-2afc-460c-8b8b-0079fc179767

---

As for `count` vs `count()` 
```js
 <Child count={count()} />
```
vs
```js
 <Child count={count} />
```

This is done this way by a design decision.  The basic, first idea is that you won't have to think if something is a function or not. Taking from your own example:

```js
const Child = (props) => {
  return <>{props.count % 2 === 0 ? <Even /> : <Odd />}</>;
};

const Child2 = (props) => {
  return <>{props.count() % 2 === 0 ? <Even /> : <Odd />}</>;
};
```

`Child` is using the value and `Child2` is using the function. That mix of different uses will make you having to look up if what you pass to the component or get, is a function or not. As in trying to do `<Child2 count={count()}/>` will error out, as `prop.count` is not a `function`.  One could possibly use a wrapper, as in `$(props.count)` which will use the value or call the function if it's a function, but then you will have to use the wrapper. 

--- 

Then there's the issue that reading a signal causes tracking (making the caller function re-run whenever the signal changes)
In your first example there is 

```js
<Child count={count()} /> // reading count signal

const Child = (props) => {
  return <>{props.count % 2 === 0 ? <Even /> : <Odd />}</>;
};
```

If `count={count()}` were really reading the signal, then `props.count` would be just a regular value and can't be reactive. Therefore, you cannot get fine-grained updates inside `Child` and instead you will have a full re-render of wherever you read `count()` and also a re-render of `Child`.

For this reason solid comes with a compiler. The compiler changes props to be getters, to "push" reading a signal down in the tree as possible.

When you write
```js
<Child count={count()} />
```
solid compiler outputs

```js
<Child count={get count() { return count() }} />
```
That way the signal is not read until you access `props.count`, pushing the signal read as deep as possible giving fine-grained updates.

```js
<Child count={get count() { return count() }} />

const Child = (props) => {
  return <>{props.count % 2 === 0 ? <Even /> : <Odd />}</>;
};
```

Granted this can be confusing, when you do 

```js
const [count, setCount] = createSignal(1)
<Child count={count()} />
```

You may expect to be the same as 

```js
const [count, setCount] = createSignal(1)
<Child count={1} />
```

but that's not the case, what happens is 

```js
const [count, setCount] = createSignal(1)
<Child count={{ get count() { return count() /* 1 */ }} />
```

Let me know if you have any doubt

**Comment by thetsp** [comment]
Yes it is working now. Thank you so much! What I understood is unlike react js, solid js reruns jsx for updates while react js runs whole component. Is my understanding correct?
